llama_finetuning.py

```diff
@@ -90,11 +95,26 @@ def main(**kwargs):
     gradient_accumulation_steps = train_config.batch_size_training // train_config.micro_batch_size
      
     # Load the pre-trained model and setup its configuration
-    model = LlamaForCausalLM.from_pretrained(
-        train_config.model_name,
-        load_in_8bit=True if train_config.quantization else None,
-        device_map="auto" if train_config.quantization else None,
-    )
+    if train_config.enable_fsdp:
+        # for FSDP, we save cpu memory by loading pretrained model on rank0 only.
+        # this avoids cpu oom when loading large models like llama 70B, in which case
+        # model alone would consume 2+TB cpu mem (70 * 4 * 8)
+        if rank == 0:
+            model = LlamaForCausalLM.from_pretrained(
+                train_config.model_name,
+                load_in_8bit=True if train_config.quantization else None,
+                device_map="auto" if train_config.quantization else None,
+            )
+        else:
+            llama_config = LlamaConfig.from_pretrained(train_config.model_name)
+            with init_empty_weights():
```

Comment:
Curious: Why do we need `init_empty_weights()` and not just `with torch.device("meta")`? What is insufficient with the latter?

Reply:
no specific reasons, originally doing it this way because the code is reading the model in HF way and specifically add `accelerate` in the requirements so thinking about using `accelerate.init_empty_weights` to match the model loading style.  Also, I remembered I hit some broadcast op issues when using with torch.device("meta") + sync_states in the past when this was introduced, let me double check it here and make the changes if looking good.

Reply:
Interesting. If I switch to `with torch.device("meta")` while keeping my current `_param_init_fn` that I borrowed from one of your test cases, then it will fail with
```
NotImplementedError: Could not run 'c10d::broadcast_' with arguments from the 'Meta' backend
```

But, if I switch both, the combination of `with torch.device("meta")` + `to_empty(recurse=False)` works well. I will make both changes then.

Reply:
addressed in https://github.com/facebookresearch/llama-recipes/pull/77/commits/1e64fc98d9296b28b7d34b11c28498f50af69e8b